src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs

```diff
@@ -56,39 +57,46 @@ public static void PopulateDefaults(PageRouteModel model, string pageRoute, stri
             result = default;
             Debug.Assert(path.StartsWith("/", StringComparison.Ordinal));
 
-            var areaEndIndex = path.IndexOf('/', startIndex: 1);
-            if (areaEndIndex == -1 || areaEndIndex == path.Length)
+            var stringTokenizer = new StringTokenizer(path, PathSeparator);
+            var enumerator = stringTokenizer.GetEnumerator();
+            var next = enumerator.MoveNext();
+            Debug.Assert(next && enumerator.Current.Length == 0, "Paths must start with leading slash");
+
+            if (!enumerator.MoveNext() || !enumerator.Current.HasValue)
             {
                 logger.UnsupportedAreaPath(razorPagesOptions, path);
                 return false;
             }
 
-            // Normalize the pages root directory so that it has a 
-            var normalizedPagesRootDirectory = razorPagesOptions.RootDirectory.TrimStart('/');
-            if (!normalizedPagesRootDirectory.EndsWith("/", StringComparison.Ordinal))
+            var areaName = enumerator.Current;
+            if (razorPagesOptions.RootDirectory != "/")
             {
-                normalizedPagesRootDirectory += "/";
-            }
-
-            if (string.Compare(path, areaEndIndex + 1, normalizedPagesRootDirectory, 0, normalizedPagesRootDirectory.Length, StringComparison.OrdinalIgnoreCase) != 0)
-            {
-                logger.UnsupportedAreaPath(razorPagesOptions, path);
-                return false;
+                Debug.Assert(razorPagesOptions.RootDirectory.StartsWith("/", StringComparison.Ordinal));
+                var rootDirectory = razorPagesOptions.RootDirectory;
+                var normalizedRootDirectory = new StringSegment(rootDirectory, 1, rootDirectory.Length - 1);
+                var rootDirectoryTokenizer = new StringTokenizer(normalizedRootDirectory, PathSeparator);
+                foreach (var segment in rootDirectoryTokenizer)
+                {
+                    if (!enumerator.MoveNext() || !segment.Equals(enumerator.Current, StringComparison.OrdinalIgnoreCase))
+                    {
+                        logger.UnsupportedAreaPath(razorPagesOptions, path);
+                        return false;
+                    }
+                }
```

Comment:
I really have mixed feelings on seeing this code written with the tokenizer. It's much harder to understand

Reply:
Fair enough. I can stick to the current code. I'll see if I can document \ clean it up further and simplify the special case of an empty root directory.